utils/frame/remote-externalities/src/lib.rs

```diff
@@ -320,72 +382,147 @@ where
-					chunk_keys.iter().map(HexDisplay::from).collect::<Vec<_>>(),
-					e
-				);
-				"batch failed."
-			})?;
+		log::info!(
+			target: LOG_TARGET,
+			"Querying a total of {} keys, splitting among {} threads, {} keys per thread",
+			keys.len(),
+			self.as_online().threads,
+			thread_chunk_size,
+		);
 
-			assert_eq!(chunk_keys.len(), values.len());
+		let mut handles = Vec::new();
+		let keys_chunked: Vec<Vec<StorageKey>> =
+			keys.chunks(thread_chunk_size).map(|s| s.into()).collect::<Vec<_>>();
 
-			for (idx, key) in chunk_keys.iter().enumerate() {
-				let maybe_value = values[idx].clone();
-				let value = maybe_value.unwrap_or_else(|| {
-					log::warn!(target: LOG_TARGET, "key {:?} had none corresponding value.", &key);
-					StorageData(vec![])
+		enum Message {
+			Terminated,
+			Batch(Vec<(Vec<u8>, Vec<u8>)>),
+		}
+		let (tx, rx) = std::sync::mpsc::channel::<Message>();
+
+		time_block!("value_download", {
+			for thread_keys in keys_chunked {
+				let thread_client = client.clone();
+				let thread_sender = tx.clone();
+				let handle = std::thread::spawn(move || {
+					let rt = tokio::runtime::Runtime::new().unwrap();
+					let mut thread_key_values = Vec::with_capacity(thread_keys.len());
+					for chunk_keys in thread_keys.chunks(DEFAULT_VALUE_DOWNLOAD_BATCH) {
+						let batch = chunk_keys
+							.iter()
+							.cloned()
+							.map(|key| ("state_getStorage", rpc_params![key, at]))
+							.collect::<Vec<_>>();
+
+						let values = rt
+							.block_on(thread_client.batch_request::<Option<StorageData>>(batch))
+							.map_err(|e| {
+								log::error!(
+									target: LOG_TARGET,
+									"failed to execute batch of {} values. Error: {:?}",
+									chunk_keys.len(),
+									e
+								);
+								"batch failed."
+							})
+							.unwrap();
+
+						let batch_kv = chunk_keys
+							.into_iter()
+							.enumerate()
+							.map(|(idx, key)| {
+								let maybe_value = values[idx].clone();
+								let value = maybe_value.unwrap_or_else(|| {
+									log::warn!(
+										target: LOG_TARGET,
+										"key {:?} had none corresponding value.",
+										&key
+									);
+									StorageData(vec![])
+								});
+								thread_key_values.push((key.clone(), value.clone()));
+								if thread_key_values.len() % (thread_keys.len() / 10).max(1) == 0 {
+									let ratio: f64 =
+										thread_key_values.len() as f64 / thread_keys.len() as f64;
+									log::debug!(
+										target: LOG_TARGET,
+										"[thread = {:?}] progress = {:.2} [{} / {}]",
+										std::thread::current().id(),
+										ratio,
+										thread_key_values.len(),
+										thread_keys.len(),
+									);
+								}
+								(key.clone().0, value.0)
+							})
+							.collect::<Vec<_>>();
+
+						// send this batch to the main thread to start inserting.
+						thread_sender.send(Message::Batch(batch_kv)).unwrap();
+					}
+
+					thread_sender.send(Message::Terminated).unwrap();
+					thread_key_values
```

Comment:
question: you've lost me here; why do we return this data? haven't we sent it all already?

Reply:
We still need to bubble it up to write it to a file, if desired. 

There's a not acknowledging that this is not so great.

I think a nice abstraction would be to pass in a `Optiom<WriteStream>` to this function. Similar to updating `pending_ext`, we also write to this. If a cache file is needed, this will be `Some(_)`, else `None`.

But I am not sure how to implement that efficiently either, so I timed the execution, and it turned out that returning this data for caching purposes is not a big deal, so I opted for this for now.